### PR TITLE
feat: Modal & Select support z-index context to manage z-index

### PR DESCRIPTION
**Pull request by kiner-tang**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [X] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

- https://github.com/ant-design/ant-design/discussions/45154

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Modal & Select Support z-index context to manage z-index |
| 🇨🇳 Chinese | 弹框和下拉框支持通过 z-index 上下文管理 z-index  |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [X] Doc is updated/provided or not needed
- [X] Demo is updated/provided or not needed
- [X] TypeScript definition is updated/provided or not needed
- [X] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at fed6250</samp>

This pull request adds support for dynamic zIndex values in the `Modal` and `Select` components. It introduces a new custom hook `useZIndex` and a new context `zIndexContext` to calculate and pass the zIndex based on the parent context and the consumer type. It also adds a new demo `nested.tsx` to show an example of using nested modals with different zIndex values.

### <samp>🤖 Generated by Copilot at fed6250</samp>

*  Add a new custom hook `useZIndex` to calculate the zIndex value for different components based on the parent context and the consumer type ([link](https://github.com/ant-design/ant-design/pull/45346/files?diff=unified&w=0#diff-7469c5684f788ea6a0ab4681cbdd7b7c8bcd12bdbc0fdd05afec1039868d2d08R1-R46))
*  Add a new module `zindexContext` to export a React context and a provider component for storing and passing the zIndex value across components ([link](https://github.com/ant-design/ant-design/pull/45346/files?diff=unified&w=0#diff-0f41c7c229ef54fe09c53f519e935ab2de4480c2baf2379bf9ac78f82eef314cR1-R11))
*  Use the `useZIndex` hook and the `ZIndexContextProvider` in the `Modal` component to wrap the `Dialog` component and pass the dynamic zIndex value to both the provider and the dialog props ([link](https://github.com/ant-design/ant-design/pull/45346/files?diff=unified&w=0#diff-a3edcd43b644d0539a67f65e68ba30d6acb77a12531a6a71d1e26a008a9d752cR10), [link](https://github.com/ant-design/ant-design/pull/45346/files?diff=unified&w=0#diff-a3edcd43b644d0539a67f65e68ba30d6acb77a12531a6a71d1e26a008a9d752cR18), [link](https://github.com/ant-design/ant-design/pull/45346/files?diff=unified&w=0#diff-a3edcd43b644d0539a67f65e68ba30d6acb77a12531a6a71d1e26a008a9d752cL116-R159))
*  Use the `useZIndex` hook in the `Select` component to pass the dynamic zIndex value to the `dropdownStyle` prop of the `RcSelect` component ([link](https://github.com/ant-design/ant-design/pull/45346/files?diff=unified&w=0#diff-cf9c6a6fcee364e93144b7afa91b8ff8b85e8bab46a1b602752c25b6c8de55d0R27), [link](https://github.com/ant-design/ant-design/pull/45346/files?diff=unified&w=0#diff-cf9c6a6fcee364e93144b7afa91b8ff8b85e8bab46a1b602752c25b6c8de55d0R244-R246), [link](https://github.com/ant-design/ant-design/pull/45346/files?diff=unified&w=0#diff-cf9c6a6fcee364e93144b7afa91b8ff8b85e8bab46a1b602752c25b6c8de55d0R273-R275))
*  Add a new demo file `nested.tsx` to show an example of using nested modals with different zIndex values using the `Modal` and `Select` components and the `useZIndex` hook ([link](https://github.com/ant-design/ant-design/pull/45346/files?diff=unified&w=0#diff-740d231aa0062e7ab64fb1c11b1a82f487d219f392ddcbf2be01b45a8538e1e6R1-R73))
*  Add a new demo file `nested.md` to display the nested modals example in the documentation ([link](https://github.com/ant-design/ant-design/pull/45346/files?diff=unified&w=0#diff-fa6c244a20ffd0c68780b5a5579d6f78f9e344b64c9b764de2301a291e5a9f79R1-R7))
*  Add a reference to the `nested.tsx` demo file in the `index.en-US.md` and `index.zh-CN.md` documentation files for the `Modal` component ([link](https://github.com/ant-design/ant-design/pull/45346/files?diff=unified&w=0#diff-0868ee605478aef1b7a7e4508dce443964207f1dd966d82be9bf7e9b16a31aa5R38), [link](https://github.com/ant-design/ant-design/pull/45346/files?diff=unified&w=0#diff-7a204621cc9425007680c097ad01151df377979692bbc09eb2796025b85043f3R39))

### CODE REVIEW

File: components/_util/hooks/useZIndex.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  | 'Menu';
+
+export const baseZIndexOffset: Record<ZIndexConsumer | ZIndexContainer, number> = {
+  Modal: 0,
+  Drawer: 0,
+  Popover: 0,
+  Popconfirm: 0,
+  Tooltip: 0,
+  Tour: 0,
+  Select: 50,
```
Comment (zombieJ) [inline components/_util/hooks/useZIndex.ts:25]: 为啥 Select 单独是个 50？ Select 和 TreeSelect 和 Cascader 不是都是类 Select 组件么？

Comment (kiner-tang) [reply components/_util/hooks/useZIndex.ts:25]: 先做了 Select 的，其他的 offset 还没有去看，如果这个方案没问题，我再把其他 offset 不上去

File: components/select/index.tsx [hunk 2]
```diff
@@ -240,6 +241,9 @@ const InternalSelect = <
     );
   }
 
+  // ====================== zIndex =========================
+  const curZIndex = useZIndex('Select');
```
Comment (zombieJ) [inline components/select/index.tsx:245]: 最好是冷启动，如果不是在 Drawer Modal 中，其实不用设置这个。

File: components/select/index.tsx [hunk 3]
```diff
@@ -266,6 +270,13 @@ const InternalSelect = <
       getPopupContainer={getPopupContainer || getContextPopupContainer}
       dropdownClassName={rcSelectRtlDropdownClassName}
       disabled={mergedDisabled}
+      dropdownStyle={
+        curZIndex == null
+          ? {}
+          : {
+              zIndex: curZIndex,
+            }
+      }
```
[the remark this answers is not in the document]
Comment (kiner-tang) [reply components/select/index.tsx:279]: 啥意思？

File: components/_util/hooks/useZIndex.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  TimePicker: 50,
+  Menu: 50,
+};
+
+export function useZIndex(consumer?: ZIndexConsumer) {
+  const [, token] = useToken();
+  const { zIndex: parentZIndex } = React.useContext(zIndexContext);
+  if (parentZIndex === null) {
+    return {
+      curZIndex: consumer ? baseZIndexOffset[consumer] : token.zIndexPopupBase,
```
Comment (zombieJ) [inline components/_util/hooks/useZIndex.ts:35]: 看起来可以简化为：
```tsx
baseZIndexOffset[consumer] ?? token.zIndexPopupBase
```

Comment (kiner-tang) [reply components/_util/hooks/useZIndex.ts:35]: done

Comment (zombieJ) [reply components/_util/hooks/useZIndex.ts]: 另外，感觉有点怪。如果是 `parentZIndex` 不存在的话，这边应该是不用返回计算值。直接交由现有的 CSS 处理就好了，不需要自己从 baseZIndexOffset 里拿

Comment (kiner-tang) [reply components/_util/hooks/useZIndex.ts]: image
这边已经做了判断了，主要是如果是容器，就算不存在 parentZIndex, 也需要把自己的 zIndex 传下去

File: components/_util/hooks/useZIndex.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  const [, token] = useToken();
+  const { zIndex: parentZIndex } = React.useContext(zIndexContext);
+  if (parentZIndex === null) {
+    return {
+      curZIndex: baseZIndexOffset[consumer!] ?? token.zIndexPopupBase,
+      parentZIndex: null,
+    };
+  }
+  let curZIndex = parentZIndex;
+  if (consumer) {
```
Comment (MadCcc) [inline components/_util/hooks/useZIndex.ts:40]: 这个 consumer 不太懂，Modal 应该也有自己偏差值，相对 Select 这些会更大

File: components/modal/Modal.tsx [hunk 6]
```diff
@@ -113,38 +115,48 @@ const Modal: React.FC<ModalProps> = (props) => {
...
+          value={{
+            zIndex: curZIndex,
           }}
-          panelRef={panelRef}
-        />
+        >
+          <Dialog
+            width={width}
+            {...restProps}
+            zIndex={parentZIndex === null ? restProps.zIndex : curZIndex}
```
Comment (zombieJ) [inline components/modal/Modal.tsx:133]: 这个逻辑太臃肿了，`useZIndex` 里应该返回全套。直接告诉你当前 `zIndex`，另外用户输入的 `zIndex` 优先级一定是最高的：
```tsx
zIndex={restProps.zIndex ?? zIndex}
```

File: components/_util/hooks/useZIndex.ts (hunk 5 above)
Comment (kiner-tang) [reply components/_util/hooks/useZIndex.ts:40]: 因为容器需要加上 `token.zIndexPopupBase`，而消费者只需要将父级的 z-index 再加上偏移量就可以了

File: components/modal/Modal.tsx (hunk 6 above)
Comment (kiner-tang) [reply components/modal/Modal.tsx:133]: 确实，我再理理

File: components/select/index.tsx [hunk 7]
```diff
@@ -266,6 +270,14 @@ const InternalSelect = <
       getPopupContainer={getPopupContainer || getContextPopupContainer}
       dropdownClassName={rcSelectRtlDropdownClassName}
       disabled={mergedDisabled}
+      dropdownStyle={
+        parentZIndex == null
```
Comment (zombieJ) [inline components/select/index.tsx:274]: 同 Drawer comment，组件不应该感知层级关系。都应该由 useZIndex 提供

File: components/_util/hooks/useZIndex.ts (hunk 5 above)
Comment (MadCcc) [reply components/_util/hooks/useZIndex.ts:40]: 容器除了 base 也有自己的偏移量

File: components/_util/hooks/useZIndex.tsx [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  TreeSelect: 50,
+  AutoComplete: 50,
+  ColorPicker: 30,
+  DatePicker: 50,
+  TimePicker: 50,
+  Menu: 50,
+};
+
+function isContainerType(type: ZIndexContainer | ZIndexConsumer): type is ZIndexContainer {
+  return Object.keys(containerBaseZIndexOffset).includes(type);
```
Comment (zombieJ) [inline components/_util/hooks/useZIndex.tsx:39]:
```tsx
type in containerBaseZIndexOffset
```

Comment (zombieJ) [reply components/_util/hooks/useZIndex.tsx:39]: 没必要转一遍~

Comment (kiner-tang) [reply components/_util/hooks/useZIndex.tsx]: done

File: components/_util/hooks/useZIndex.tsx [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+};
+
+function isContainerType(type: ZIndexContainer | ZIndexConsumer): type is ZIndexContainer {
+  return type in containerBaseZIndexOffset;
+}
+
+export function useZIndex(componentType: ZIndexContainer | ZIndexConsumer) {
+  const [, token] = useToken();
+  const { zIndex: parentZIndex } = React.useContext(zIndexContext);
+  const isContainer = isContainerType(componentType!);
```
Comment (zombieJ) [inline components/_util/hooks/useZIndex.tsx:45]: 上面定义 `type?: ZIndexContainer | ZIndexConsumer` 这边就不用强行用 `!` 了

File: components/modal/Modal.tsx [hunk 10]
```diff
@@ -113,38 +115,48 @@ const Modal: React.FC<ModalProps> = (props) => {
...
-            wrapper: wrapClassNameExtended,
-            ...modal?.classNames,
-            ...modalClassNames,
-          }}
-          styles={{
-            ...modal?.styles,
-            ...modalStyles,
+        <ZIndexContextProvider
+          value={{
+            zIndex: zIndex ?? containerBaseZIndex,
```
Comment (MadCcc) [inline components/modal/Modal.tsx:127]: parentZIndex 是 null 的时候这里 zIndex 是 null，而 containerBaseZIndex 的计算逻辑和 zIndex 一模一样，是不是没必要返回两个值在这里做判断？

File: components/_util/hooks/useZIndex.tsx (hunk 9 above)
Comment (kiner-tang) [reply components/_util/hooks/useZIndex.tsx:45]: 确实，遗留下来的，我去掉

Comment (kiner-tang) [reply components/_util/hooks/useZIndex.tsx]: done

File: components/_util/zindexContext.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+
+export type ZIndexContext = {
+  zIndex: number | null;
+};
+
+export const zIndexContext = React.createContext<ZIndexContext>({
+  zIndex: null,
```
Comment (zombieJ) [inline components/_util/zindexContext.ts:8]: 如果只有单个值，这边可以直接 number， 不用包一个对象

File: components/_util/hooks/useZIndex.tsx [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  let containerBaseZIndex = 0;
+  if (isContainer) {
+    containerBaseZIndex = token.zIndexPopupBase + containerBaseZIndexOffset[componentType];
+  }
+  if (parentZIndex === null) {
+    return {
+      zIndex: null,
+      containerBaseZIndex,
+    };
+  }
```
Comment (MadCcc) [inline components/_util/hooks/useZIndex.tsx:55]: 当 parentZIndex 是 null 的时候组件就不需要对这个 hook 做出响应了，直接用样式里的 zIndex 就行了

File: components/modal/Modal.tsx (hunk 10 above)
Comment (kiner-tang) [reply components/modal/Modal.tsx:127]: 因为要确保组件如果没有嵌套在弹出容器中的时候，不在 style 加额外的 z-index，所以，parentZIndex 为 null 的时候，zIndex 也要返回一个 null 作为特殊值跳过相关逻辑。但作为 Provider，有需要讲当前组件的 z-index 传下去给子组件消费，感觉没办法省

Comment (zombieJ) [inline components/modal/Modal.tsx:127]: 这里漏了 Modal 本身的 zIndex prop

File: components/_util/zindexContext.ts (hunk 11 above)
Comment (kiner-tang) [reply components/_util/zindexContext.ts:8]: 不知道后续会不会扩展，类似 SizeContext 之类的，目前倒是可以改成 number
image

File: components/select/index.tsx [hunk 13]
```diff
@@ -266,6 +270,10 @@ const InternalSelect = <
       getPopupContainer={getPopupContainer || getContextPopupContainer}
       dropdownClassName={rcSelectRtlDropdownClassName}
       disabled={mergedDisabled}
+      dropdownStyle={{
+        ...props?.dropdownStyle,
+        zIndex: props.dropdownStyle?.zIndex ?? zIndex ?? undefined,
```
Comment (zombieJ) [inline components/select/index.tsx:275]: 后面那个 `?? undefined` 看起来是多余的

Comment (kiner-tang) [reply components/select/index.tsx:275]: image

File: components/modal/Modal.tsx (hunk 10 above)
Comment (kiner-tang) [reply components/modal/Modal.tsx]: done

File: components/_util/zindexContext.ts (hunk 11 above)
Comment (kiner-tang) [reply components/_util/zindexContext.ts]: done

File: components/_util/hooks/useZIndex.tsx (hunk 12 above)
Comment (kiner-tang) [reply components/_util/hooks/useZIndex.tsx:55]: 因为要确保组件如果没有嵌套在弹出容器中的时候，不在 style 加额外的 z-index，所以，parentZIndex 为 null 的时候，zIndex 也要返回一个 null 作为特殊值跳过相关逻辑。但作为 Provider，有需要讲当前组件的 z-index 传下去给子组件消费，感觉没办法省

Comment (MadCcc) [reply components/_util/hooks/useZIndex.tsx:55]: 应该去掉多余的数值变量，用一个 boolean 值来表示是否需要启用 zIndex 配置。
同样的计算逻辑重复写了两遍，会让代码更难维护

Comment (MadCcc) [reply components/_util/hooks/useZIndex.tsx:55]: 一点实现上的建议，看看合不合适。

ZIndexProvider 目前只会在 Modal 这种组件中使用，并且逻辑都是一致的，可以封装一下

这样在 Modal 里应该是这样的
```
// zIndex 可能是 undefined
const { zIndex } = useZIndex();

<ZIndexProvider>
  <Modal zIndex={zIndex} />
</ZIndexProvider>
```

ZIndexProvider 实现：
```
() => {
  const { containerZIndex } = useZIndex();

  return (
    <ZIndexContext.Provider value={{ zIndex: containerZIndex }}>
      {children}
    </ZIndexContext.Provider>
}
```

File: components/_util/hooks/useZIndex.tsx [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+  return type in containerBaseZIndexOffset;
+}
+
+export function useZIndex(componentType: ZIndexContainer | ZIndexConsumer) {
+  const [, token] = useToken();
+  const parentZIndex = React.useContext(zIndexContext);
+  const isContainer = isContainerType(componentType);
+  let containerBaseZIndex = 0;
+  if (isContainer) {
+    containerBaseZIndex = token.zIndexPopupBase + containerBaseZIndexOffset[componentType];
```
Comment (MadCcc) [inline components/_util/hooks/useZIndex.tsx:48]: 这个数值应该就是计算出来的组件 zIndex，所以没必要和 zIndex 分开计算吧。
应该这样返回就够了
```
return {
  zIndex: parentZIndex ? zIndex : undefiend,
  containerZIndex: zIndex,
}
```

**Comment by kiner-tang** [comment]
@zombieJ @MadCcc 再帮忙看看，如果没问题我就先合过去，然后我把其他组件按照这个方案改一下。

File: components/_util/hooks/useZIndex.tsx [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  const isContainer = isContainerType(componentType);
+  let zIndex = parentZIndex ?? 0;
+  if (isContainer) {
+    zIndex += token.zIndexPopupBase + containerBaseZIndexOffset[componentType];
+  } else {
+    zIndex += consumerBaseZIndexOffset[componentType];
+  }
+  return {
+    zIndex: parentZIndex ? zIndex : undefined,
+    containerZIndex: zIndex,
```
Comment (zombieJ) [inline components/_util/hooks/useZIndex.tsx:54]: zIndex 和 containerZIndex 差异已经不大了，感觉只需要提供一个就好了。自己当前的 zIndex 也是扔给下一层的起始 zIndex

Comment (kiner-tang) [reply components/_util/hooks/useZIndex.tsx:54]: 这个不好去掉吧，因为在没有父容器的时候，组件就不需要更新 style 里的 zIndex，只有有父容器的时候才要

Comment (kiner-tang) [reply components/_util/hooks/useZIndex.tsx:54]: done

File: components/_util/zindexContext.ts [hunk 16]
```diff
@@ -0,0 +1,7 @@
+import React from 'react';
+
+export type ZIndexContext = number | undefined;
```
Comment (zombieJ) [inline components/_util/zindexContext.ts:3]: 单独数字不用 export 了，直接 inline 写就好

File: components/_util/zindexContext.ts [hunk 17]
```diff
@@ -0,0 +1,7 @@
...
+
+export const zIndexContext = React.createContext<ZIndexContext>(undefined);
+
+export const ZIndexContextProvider = zIndexContext.Provider;
```
Comment (zombieJ) [inline components/_util/zindexContext.ts:7]: 消费的地方直接 Context.Provider 就好，一般提供 XXXProvider 都是有额外逻辑封装才需要。

File: components/select/index.tsx [hunk 18]
```diff
@@ -240,6 +241,9 @@ const InternalSelect = <
     );
   }
 
+  // ====================== zIndex =========================
+  const { zIndex } = useZIndex('Select', props.dropdownStyle?.zIndex as number);
```
Comment (zombieJ) [inline components/select/index.tsx:245]: 诶？这个居然不是 popupStyle 么？

Comment (kiner-tang) [reply components/select/index.tsx:245]: Select 里确实还是 dropdownStyle ,之后可以单独开个 PR 改成 popupStyle

File: components/_util/zindexContext.ts (hunk 17 above)
Comment (kiner-tang) [reply components/_util/zindexContext.ts]: done

File: components/_util/zindexContext.ts (hunk 16 above)
Comment (kiner-tang) [reply components/_util/zindexContext.ts]: done

File: components/modal/demo/nested.tsx [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+    label: 'Option 1',
+    value: '1',
+  },
+  {
+    label: 'Option 2',
+    value: '2',
+  },
+];
+
+const App: React.FC = () => {
```
Comment (zombieJ) [inline components/modal/demo/nested.tsx:15]: 这个 demo 可以简化成 Switch 组件，打开后把 3 个 Modal 都打开。并且禁用 mask，通过高度把 3 个 Select 都展示出来。

Comment (kiner-tang) [reply components/modal/demo/nested.tsx:15]: 好

File: components/modal/demo/basic.tsx [hunk 20]
```diff
@@ -21,7 +21,13 @@ const App: React.FC = () => {
       <Button type="primary" onClick={showModal}>
         Open Modal
       </Button>
-      <Modal title="Basic Modal" open={isModalOpen} onOk={handleOk} onCancel={handleCancel}>
+      <Modal
+        title="Basic Modal"
+        zIndex={9998}
```
Comment (zombieJ) [inline components/modal/demo/basic.tsx:26]: 这个没清理

**Review by zombieJ — APPROVED** [review]

**Comment by kiner-tang** [comment]
> 回退到5.10.3了，期待下个版本。

目前已知的 z-index 已经在最新版本修复了，这边是还遇到了其他问题吗？

**Comment by h1orz** [comment]
> > 回退到5.10.3了，期待下个版本。
> 
> 目前已知的 z-index 已经在最新版本修复了，这边是还遇到了其他问题吗？

之前的版本z-index+50，现在+200，影响了业务代码，加上Select、Dropdown的层级也受了影响，刚好从5.11.0那几天升级，心脏遭不住，等等再升级。

**Pull request merged**